testcases/kernel/syscalls/inotify/inotify08.c

```diff
@@ -0,0 +1,239 @@
+
+void verify_inotify(void)
+{
+	int test_cnt = 0;
+
+	/*
+	 * generate sequence of events
+	 */
+	SAFE_CHMOD(FILE_PATH, 0644);
+	event_set[test_cnt].mask = IN_ATTRIB;
+	test_cnt++;
+
+	SAFE_FILE_PRINTF(FILE_PATH, "1");
+
+	event_set[test_cnt].mask = IN_OPEN;
+	test_cnt++;
+
+	event_set[test_cnt].mask = IN_CLOSE_WRITE;
+	test_cnt++;
+
+	/* Make sure events on upper/lower do not show in overlay watch */
+	SAFE_TOUCH("lower/"FILE_NAME, 0644, NULL);
+	SAFE_TOUCH("upper/"FILE_NAME, 0644, NULL);
+
+	int len, i = 0, test_num = 0;
+	if ((len = read(fd_notify, event_buf, EVENT_BUF_LEN)) == -1) {
+		tst_brk(TBROK | TERRNO,
+			"read(%d, buf, %zu) failed",
+			fd_notify, EVENT_BUF_LEN);
+
+	}
+
+	while (i < len) {
+		struct inotify_event *event;
+		event = (struct inotify_event *)&event_buf[i];
+		if (test_num >= test_cnt) {
+			tst_res(TFAIL,
+				"get unnecessary event: "
+				"wd=%d mask=%08x cookie=%-5u len=%-2u "
+				"name=\"%.*s\"", event->wd, event->mask,
+				event->cookie, event->len, event->len,
+				event->name);
+		} else if (event_set[test_num].mask == event->mask &&
+			   !event->len) {
+			tst_res(TPASS,
+				"get event: wd=%d mask=%08x "
+				"cookie=%-5u len=%-2u",
+				event->wd, event->mask,
+				event->cookie, event->len);
+		} else {
+			tst_res(TFAIL, "get event: wd=%d mask=%08x "
+				"(expected %x) cookie=%-5u len=%-2u "
+				"name=\"%.*s\" (expected \"\")",
+				event->wd, event->mask,
+				event_set[test_num].mask,
+				event->cookie, event->len, event->len,
+				event->name);
+		}
+		test_num++;
+		i += EVENT_SIZE + event->len;
+	}
+
+	for (; test_num < test_cnt; test_num++) {
+		tst_res(TFAIL, "didn't get event: mask=%x ",
+			event_set[test_num].mask);
+	}
+}
+
+static void setup(void)
+{
+	struct stat buf;
+	int ret;
+
+	/* Setup an overlay mount with lower file */
+	SAFE_MKDIR("lower", 0755);
+	SAFE_TOUCH("lower/"FILE_NAME, 0644, NULL);
+	SAFE_MKDIR("upper", 0755);
+	SAFE_MKDIR("work", 0755);
+	SAFE_MKDIR(OVL_MNT, 0755);
+	ret = mount("overlay", OVL_MNT, "overlay", 0,
+		    "lowerdir=lower,upperdir=upper,workdir=work");
+	if (ret < 0) {
+		if (errno == ENODEV) {
+			tst_brk(TCONF,
+				"overlayfs is not configured in this kernel.");
+		} else {
+			tst_brk(TBROK | TERRNO,
+				"overlayfs mount failed");
+		}
+	}
+	ovl_mounted = 1;
+
+	fd_notify = myinotify_init1(O_NONBLOCK);
+	if (fd_notify < 0) {
+		if (errno == ENOSYS) {
+			tst_brk(TCONF,
+				"inotify is not configured in this kernel.");
+		} else {
+			tst_brk(TBROK | TERRNO,
+				"inotify_init () failed");
+		}
+	}
+
+	/* Setup a watch on an overlayfs lower file */
+	if ((wd = myinotify_add_watch(fd_notify, FILE_PATH,
+				IN_ATTRIB | IN_OPEN | IN_CLOSE_WRITE)) < 0) {
+		tst_brk(TBROK | TERRNO,
+			"inotify_add_watch (%d, " FILE_PATH ", "
+			"IN_ATTRIB | IN_OPEN | IN_CLOSE_WRITE) failed",
+			fd_notify);
+		reap_wd = 1;
+	};
+
+	SAFE_STAT(FILE_PATH, &buf);
+	tst_res(TINFO, FILE_PATH " ino=%lu, dev=%u:%u", buf.st_ino,
+			major(buf.st_dev), minor(buf.st_dev));
+
```

Comment:
I've also got a header deprecation warnings on newer systems for the minor() and major():

```
inotify08.c:211:13: warning: In the GNU C Library, "minor" is defined
 by <sys/sysmacros.h>. For historical compatibility, it is
 currently defined by <sys/types.h> as well, but we plan to
 remove this soon. To use "minor", include <sys/sysmacros.h>
 directly. If you did not intend to use a system-defined macro
 "minor", you should undefine it after including <sys/types.h>.
```

So we should add #include <sys/sysmacros.h> to the includes here.